src/torchcodec/decoders/_core/VideoDecoder.cpp

```diff
@@ -782,8 +825,43 @@ VideoDecoder::DecodedOutput VideoDecoder::convertAVFrameToDecodedOutput(
   output.durationSeconds = ptsToSeconds(
       getDuration(frame), formatContext_->streams[streamIndex]->time_base);
   if (output.streamType == AVMEDIA_TYPE_VIDEO) {
-    output.frame =
-        convertFrameToTensorUsingFilterGraph(streamIndex, frame.get());
+    if (streamInfo.filterState.sourceContext == nullptr) {
+      auto start = std::chrono::high_resolution_clock::now();
+      int width = streamInfo.options.width.value_or(frame->width);
+      int height = streamInfo.options.height.value_or(frame->height);
+      torch::Tensor tensor = torch::empty(
+          {height, width, 3}, torch::TensorOptions().dtype({torch::kUInt8}));
+      auto allocateDone = std::chrono::high_resolution_clock::now();
+      rawOutput.data = tensor.data_ptr<uint8_t>();
+      convertFrameToBufferUsingSwsScale(rawOutput);
+      auto convertDone = std::chrono::high_resolution_clock::now();
+      auto startToAllocateDone =
+          std::chrono::duration_cast<std::chrono::microseconds>(
+              allocateDone - start);
+      auto allocateToConvertDone =
+          std::chrono::duration_cast<std::chrono::microseconds>(
+              convertDone - allocateDone);
+      auto total = std::chrono::duration_cast<std::chrono::microseconds>(
```

Comment:
I will refactor this in a different diff. It's probably not that bad. A self-profiling decoder is great imo

Reply:
Things tend to look more obvious in the eyes of the code author :) 

In this block there are 12 statements, and 5 of those are dedicated to the core logic. The 7 remaining ones are about benchmarking.

Do we expect such benchmark info to be forever useful, or is this more of a one-time thing?

Before leaving this out for a follow-up diff, I would suggest to protect the benchmarking logic within `#ifdef` blocks, so that they can easily be collapsed on the IDE to ease code review / reading.

Reply:
I think benchmark would be useful forever.

I think #ifdefs may make it even harder to read.

Later on I plan to add a timing class that will have its own code #ifdefs commented out for non-developer builds.

Should I really add #ifdefs in this diff? Leaving this unresolved.

Reply:
I actually agree with @NicolasHug that this code is now hard to read. My reasons, which some are the same as what Nicolas said:

1. The timing code takes up more lines than the actual logic.
2. Because we're using `auto` as the type, I have to actually read each timing line to realize it's a timing line. I can't just skim over them and focus on the actual logic. (Unfortunately, [explicitly stating the return type](https://fburl.com/r4137aac) would be even noisier.)
3. We're not using any new lines to group code together. "Paragraphs" of code would help me figure out what I can ignore as timing, as I expect a paragraph of code to have a start and stop timing call.

This timing code is very granular. I understand that in this PR, the amount of time just a few statements took was key to getting a big perf win, but I don't think that necessarily means we should keep it all the time. I consider it similar to putting in print statements when debugging. Some print statements graduate to actual logging, but most don't, and we delete most debugging print statements at commit time.

I would prefer either:

1. We remove the timing code in this PR, and wait until we have a readable abstraction to commit it. _OR_:
2. We limit the timing code to function boundaries. Yes, that will not be enough to fully diagnose perf problems, but it's enough to know where to start digging.

Reply:
Disagree and commit. Removed those lines